File: src/products/viteplus-monorepo/packages/mythrad-client-core/test/wire.test.ts
// Golden vectors are written out byte by byte, derived by hand from the
// layouts in the design doc rather than from this codec's own output. A
// round-trip test only proves the encoder and decoder agree with each
// other; these prove they agree with the Go server.

import { describe, expect, it } from "vitest";
import {
  CHECK_BYTES,
  DatagramKind,
  FrameKind,
  PROTO_VERSION,
  Role,
  ServerFrameKind,
  VARINT_MAX,
  VERDICT_BYTES,
  WireError,
  decodeCheck,
  decodeClientFrame,
  decodeEvent,
  decodeHello,
  decodeIntent,
  decodeReject,
  decodeResync,
  decodeServerFrame,
  decodeSnapshot,
  decodeVerdict,
  decodeWelcome,
  encodeCheck,
  encodeEvent,
  encodeHello,
  encodeIntent,
  encodeReject,
  encodeResync,
  encodeSnapshot,
  encodeVerdict,
  encodeWelcome,
  moduleHex,
  moduleWordHex,
  hex64,
  moduleWord,
  readVarint,
  varintLen,
  writeVarint,
} from "../src/wire.ts";

/** Parses a whitespace-separated hex dump into bytes, so goldens read like a spec. */
function h(dump: string): Uint8Array {
  const parts = dump.trim().split(/\s+/);
  return Uint8Array.from(parts.map((p) => parseInt(p, 16)));
}

function dv(b: Uint8Array): DataView {
  return new DataView(b.buffer, b.byteOffset, b.byteLength);
}

/** Splits a frame into its varint prefix, kind byte, and body. */
function split(frame: Uint8Array): { kind: number; body: Uint8Array } {
  const header = readVarint(dv(frame), 0);
  if (header === null) throw new Error("truncated varint");
  const start = header.next;
  const end = start + Number(header.value);
  expect(end).toBe(frame.length);
  return { kind: frame[start]!, body: frame.slice(start + 1, end) };
}

describe("QUIC varint (RFC 9000 §16)", () => {
  // The four worked examples from RFC 9000 Appendix A.1.
  const vectors: [string, bigint][] = [
    ["c2 19 7c 5e ff 14 e8 8c", 151288809941952652n],
    ["9d 7f 3e 7d", 494878333n],
    ["7b bd", 15293n],
    ["25", 37n],
  ];

  it.each(vectors)("decodes %s", (dump, want) => {
    const b = h(dump);
    const got = readVarint(dv(b), 0);
    expect(got).not.toBeNull();
    expect(got!.value).toBe(want);
    expect(got!.next).toBe(b.length);
  });

  it.each(vectors)("re-encodes %s canonically", (dump, want) => {
    const out = new Uint8Array(8);
    const next = writeVarint(dv(out), 0, want);
    expect([...out.subarray(0, next)]).toEqual([...h(dump)]);
  });

  it("decodes the RFC's non-canonical two-byte 37", () => {
    // The RFC allows a longer encoding than necessary; we must read it
    // even though we never write it.
    const got = readVarint(dv(h("40 25")), 0);
    expect(got).toEqual({ value: 37n, next: 2 });
    expect(varintLen(37)).toBe(1);
  });

  it("picks the shortest form at every boundary", () => {
    expect(varintLen(63)).toBe(1);
    expect(varintLen(64)).toBe(2);
    expect(varintLen(16383)).toBe(2);
    expect(varintLen(16384)).toBe(4);
    expect(varintLen(2 ** 30 - 1)).toBe(4);
    expect(varintLen(2 ** 30)).toBe(8);
    expect(varintLen(VARINT_MAX)).toBe(8);
  });

  it("refuses values a varint cannot carry", () => {
    expect(() => varintLen(VARINT_MAX + 1n)).toThrow(RangeError);
    expect(() => varintLen(-1)).toThrow(RangeError);
  });

  it("returns null when the encoded length is not all there yet", () => {
    const b = h("c2 19 7c");
    expect(readVarint(dv(b), 0)).toBeNull();
    expect(readVarint(dv(h("25")), 1)).toBeNull();
  });
});

describe("client frame goldens", () => {
  it("hello", () => {
    // varint 25 | kind 1 | proto=4 | since_seq=-1 | since_tick=0 |
    // ticket_len=4 | ticket=deadbeef
    const golden = h(`
      19 01
      04 00
      ff ff ff ff ff ff ff ff
      00 00 00 00 00 00 00 00
      04 00
      de ad be ef
    `);
    const value = {
      proto: PROTO_VERSION,
      sinceSeq: -1n,
      sinceTick: 0n,
      ticket: h("de ad be ef"),
    };
    expect([...encodeHello(value)]).toEqual([...golden]);
    const { kind, body } = split(golden);
    expect(kind).toBe(FrameKind.hello);
    expect(decodeHello(body)).toEqual(value);
  });

  it("intent: move_to for dog 1122334455667788 to node 300", () => {
    // varint 23 | kind 2 | id=0x0000000100000002 | kind=4 | plen=10 |
    // p = dog u64 || node u16
    const golden = h(`
      17 02
      02 00 00 00 01 00 00 00
      04 00
      0a 00
      88 77 66 55 44 33 22 11
      2c 01
    `);
    const value = {
      id: 0x0000_0001_0000_0002n,
      kind: 4,
      p: h("88 77 66 55 44 33 22 11 2c 01"),
    };
    expect([...encodeIntent(value)]).toEqual([...golden]);
    const { kind, body } = split(golden);
    expect(kind).toBe(FrameKind.intent);
    expect(decodeIntent(body)).toEqual(value);
  });

  it("resync", () => {
    const golden = h("09 03 2a 00 00 00 00 00 00 00");
    expect([...encodeResync({ haveSeq: 42n })]).toEqual([...golden]);
    const { kind, body } = split(golden);
    expect(kind).toBe(FrameKind.resync);
    expect(decodeResync(body)).toEqual({ haveSeq: 42n });
  });

  it("an empty intent payload still carries its plen", () => {
    const golden = h("0d 02 01 00 00 00 00 00 00 00 03 00 00 00");
    expect([...encodeIntent({ id: 1n, kind: 3, p: new Uint8Array(0) })]).toEqual([...golden]);
  });
});

describe("server frame goldens", () => {
  it("welcome", () => {
    // varint 46 | kind 16 | epoch=7 | seq=100 | tick=2400 | hz=24 |
    // role=1 (player) | terrain=0x0123456789abcdef | park_len=11 |
    // "park-mythra"
    const golden = h(`
      2e 10
      07 00 00 00
      64 00 00 00 00 00 00 00
      60 09 00 00 00 00 00 00
      18 00 00 00
      01
      ef cd ab 89 67 45 23 01
      0b
      70 61 72 6b 2d 6d 79 74 68 72 61
    `);
    const value = {
      epoch: 7,
      seq: 100n,
      tick: 2400n,
      hz: 24,
      role: Role.player,
      terrain: 0x0123_4567_89ab_cdefn,
      park: "park-mythra",
    };
    expect([...encodeWelcome(value)]).toEqual([...golden]);
    const { kind, body } = split(golden);
    expect(kind).toBe(ServerFrameKind.welcome);
    expect(decodeWelcome(body)).toEqual(value);
    expect(decodeServerFrame(kind, body)).toEqual({ kind: "welcome", value });
  });

  it("event: boost_set at seq 101", () => {
    // varint 38 | kind 17 | seq=101 | tick=2401 | kind=8 |
    // intent=0x0000000100000002 | plen=9 | p = dog u64 || on
    const golden = h(`
      26 11
      65 00 00 00 00 00 00 00
      61 09 00 00 00 00 00 00
      08 00
      02 00 00 00 01 00 00 00
      09 00
      88 77 66 55 44 33 22 11 01
    `);
    const value = {
      seq: 101n,
      tick: 2401n,
      kind: 8,
      intent: 0x0000_0001_0000_0002n,
      p: h("88 77 66 55 44 33 22 11 01"),
    };
    expect([...encodeEvent(value)]).toEqual([...golden]);
    const { kind, body } = split(golden);
    expect(kind).toBe(ServerFrameKind.event);
    expect(decodeEvent(body)).toEqual(value);
  });

  it("reject", () => {
    // varint 13 | kind 18 | intent | reason=3 (ERR_ABSENT)
    const golden = h("0d 12 02 00 00 00 01 00 00 00 03 00 00 00");
    const value = { intent: 0x0000_0001_0000_0002n, reason: 3 };
    expect([...encodeReject(value)]).toEqual([...golden]);
    const { kind, body } = split(golden);
    expect(kind).toBe(ServerFrameKind.reject);
    expect(decodeReject(body)).toEqual(value);
  });

  it("snapshot", () => {
    // varint 44 | kind 19 | seq=100 | tick=2400 | epoch=7 |
    // wh=0xfeedfacecafebeef | terrain=0x0123456789abcdef | zlen=3 | z
    const golden = h(`
      2c 13
      64 00 00 00 00 00 00 00
      60 09 00 00 00 00 00 00
      07 00 00 00
      ef be fe ca ce fa ed fe
      ef cd ab 89 67 45 23 01
      03 00 00 00
      01 02 03
    `);
    const value = {
      seq: 100n,
      tick: 2400n,
      epoch: 7,
      wh: 0xfeed_face_cafe_beefn,
      terrain: 0x0123_4567_89ab_cdefn,
      z: h("01 02 03"),
    };
    expect([...encodeSnapshot(value)]).toEqual([...golden]);
    const { kind, body } = split(golden);
    expect(kind).toBe(ServerFrameKind.snapshot);
    expect(decodeSnapshot(body)).toEqual(value);
  });

  it("a payload past 63 bytes takes a two-byte varint prefix", () => {
    // 40 + 24 = 64 bytes of payload, +1 kind = 65, which no longer fits
    // the 6-bit form: 0x4041 is 65 with the two-byte tag.
    const frame = encodeSnapshot({
      seq: 0n,
      tick: 0n,
      epoch: 0,
      wh: 0n,
      terrain: 0n,
      z: new Uint8Array(24),
    });
    expect(frame.length).toBe(2 + 65);
    expect([...frame.subarray(0, 2)]).toEqual([0x40, 0x41]);
  });
});

describe("datagram goldens", () => {
  it("check", () => {
    // kind 1 | tick=2400 | wh=0xfeedfacecafebeef | ct_ms=1000000
    const golden = h(`
      01
      60 09 00 00 00 00 00 00
      ef be fe ca ce fa ed fe
      40 42 0f 00 00 00 00 00
    `);
    expect(golden.length).toBe(CHECK_BYTES);
    const value = { tick: 2400n, wh: 0xfeed_face_cafe_beefn, ctMs: 1_000_000n };
    expect([...encodeCheck(value)]).toEqual([...golden]);
    expect(decodeCheck(golden)).toEqual(value);
    expect(golden[0]).toBe(DatagramKind.check);
  });

  it("verdict: known and ok", () => {
    // kind 2 | tick=2400 | now=1000050 | ct_ms=1000000 | flags=3 |
    // cw=0x11223344 | pw=0xaabbccdd
    const golden = h(`
      02
      60 09 00 00 00 00 00 00
      72 42 0f 00 00 00 00 00
      40 42 0f 00 00 00 00 00
      03
      44 33 22 11
      dd cc bb aa
    `);
    // The design doc annotates verdict as 38 B; its own field list sums to
    // 34, and the field list is what encodes and decodes.
    expect(golden.length).toBe(VERDICT_BYTES);
    expect(VERDICT_BYTES).toBe(34);
    const value = {
      tick: 2400n,
      now: 1_000_050n,
      ctMs: 1_000_000n,
      known: true,
      ok: true,
      cw: h("44 33 22 11"),
      pw: h("dd cc bb aa"),
    };
    expect([...encodeVerdict(value)]).toEqual([...golden]);
    expect(decodeVerdict(golden)).toEqual(value);
  });

  it("an unknown tick reports not-ok regardless of the ok bit", () => {
    // flags 0: neither known nor ok. A server that sets ok without known
    // is still reported as unknown — `ok` is only defined when `known`.
    const raw = encodeVerdict({
      tick: 1n,
      now: 2n,
      ctMs: 3n,
      known: false,
      ok: true,
      cw: new Uint8Array(4),
      pw: new Uint8Array(4),
    });
    expect(raw[25]).toBe(0);
    const back = decodeVerdict(raw);
    expect(back.known).toBe(false);
    expect(back.ok).toBe(false);
  });

  it("a mismatch is known but not ok", () => {
    const raw = encodeVerdict({
      tick: 9n,
      now: 0n,
      ctMs: 0n,
      known: true,
      ok: false,
      cw: new Uint8Array(4),
      pw: new Uint8Array(4),
    });
    expect(raw[25]).toBe(1);
    expect(decodeVerdict(raw)).toMatchObject({ known: true, ok: false });
  });

  it("refuses a datagram of the wrong length or kind", () => {
    expect(() => decodeCheck(new Uint8Array(24))).toThrow(WireError);
    expect(() => decodeVerdict(new Uint8Array(38))).toThrow(WireError);
    const check = encodeCheck({ tick: 0n, wh: 0n, ctMs: 0n });
    expect(() => decodeVerdict(new Uint8Array(VERDICT_BYTES))).toThrow(/not a verdict/);
    check[0] = 9;
    expect(() => decodeCheck(check)).toThrow(/not a check/);
  });
});

describe("module words and blob ids", () => {
  it("carries the module bytes verbatim and loads them LE for the ABI", () => {
    // cw/pw are opaque [4]u8: /wt-info's display string is those bytes
    // hexed left-to-right, with no endianness involved.
    const wire = h("de ad be ef");
    expect(moduleHex(wire)).toBe("deadbeef");
    // Only the wasm ABI wants an integer, and that integer is the LE load.
    expect(moduleWord(wire)).toBe(0xefbe_adde);
    expect(moduleWordHex(moduleWord(wire))).toBe("deadbeef");
  });

  it("renders a terrain blob id the way /terrain/<hex> wants it", () => {
    // Unlike the module word, park.go formats this one with %016x, so the
    // nibble order is the plain big-endian rendering of the u64.
    expect(hex64(0x0123_4567_89ab_cdefn)).toBe("0123456789abcdef");
    expect(hex64(1n)).toBe("0000000000000001");
    expect(hex64(-1n)).toBe("ffffffffffffffff");
  });
});

describe("round trips", () => {
  it("survives every payload length a u16 can name", () => {
    for (const plen of [0, 1, 63, 64, 255, 256, 16383, 16384, 65535]) {
      const p = new Uint8Array(plen).map((_, i) => i & 0xff);
      const frame = encodeIntent({ id: 7n, kind: 2, p });
      const { kind, body } = split(frame);
      expect(decodeClientFrame(kind, body)).toEqual({
        kind: "intent",
        value: { id: 7n, kind: 2, p },
      });
    }
  });

  it("keeps seq signed and tick unsigned at their extremes", () => {
    const value = {
      seq: -(2n ** 63n),
      tick: 2n ** 64n - 1n,
      kind: 0xffff,
      intent: 2n ** 64n - 1n,
      p: new Uint8Array(0),
    };
    const { kind, body } = split(encodeEvent(value));
    expect(decodeServerFrame(kind, body)).toEqual({ kind: "event", value });
  });

  it("refuses a frame kind from another protocol version", () => {
    expect(() => decodeServerFrame(20, new Uint8Array(0))).toThrow(WireError);
    expect(() => decodeClientFrame(16, new Uint8Array(0))).toThrow(WireError);
  });

  it("refuses a body longer than the fields it declares", () => {
    // The other half of "declared length equals bytes written": an
    // encoder that promised one length and wrote more shows up here as a
    // frame with trailing bytes. The Go decoders reject those too, so a
    // bug on either side surfaces at the first frame instead of drifting.
    const { kind, body } = split(
      encodeEvent({
        seq: 1n,
        tick: 2n,
        kind: 3,
        intent: 4n,
        p: Uint8Array.of(5),
      }),
    );
    const padded = new Uint8Array(body.length + 1);
    padded.set(body);
    expect(() => decodeServerFrame(kind, padded)).toThrow(/body is 30 bytes, want 29/);
    expect(() => decodeReject(new Uint8Array(13))).toThrow(WireError);
    expect(() => decodeResync(new Uint8Array(9))).toThrow(WireError);
  });

  it("refuses a body that is shorter than the fields it declares", () => {
    const good = encodeWelcome({
      epoch: 1,
      seq: 1n,
      tick: 1n,
      hz: 24,
      role: Role.spectator,
      terrain: 1n,
      park: "park-mythra",
    });
    const { body } = split(good);
    expect(() => decodeWelcome(body.slice(0, body.length - 1))).toThrow(WireError);
    expect(() => decodeWelcome(body.slice(0, 20))).toThrow(WireError);
  });
});

File: src/services/mythrad/wire/wire_test.go
package wire

import (
	"bufio"
	"bytes"
	"encoding/hex"
	"errors"
	"io"
	"reflect"
	"strings"
	"testing"
)

// The golden vectors below are the cross-language contract: the Rust
// session core and the TS host must emit these exact bytes for these exact
// inputs. They are written out by hand from the protocol layout, not
// captured from this encoder, so a change here that "fixes" a test is a
// protocol change and needs the other two implementations moved with it.

// Sim event kinds referenced by the vectors; the authority owns the full
// list.
const (
	evJoin     = 1
	evCheckIn  = 3
	evMoveTo   = 4
	evBoostSet = 8
)

func mustHex(t *testing.T, s string) []byte {
	t.Helper()
	b, err := hex.DecodeString(strings.ReplaceAll(s, " ", ""))
	if err != nil {
		t.Fatalf("bad golden vector: %v", err)
	}
	return b
}

func TestVarintRFC9000(t *testing.T) {
	// Appendix A.1's sample encodings, plus the non-minimal two-byte form
	// of 37 that a decoder must still accept.
	cases := []struct {
		enc      string
		v        uint64
		minimal  bool
		encLenIs int
	}{
		{"c2197c5eff14e88c", 151288809941952652, true, 8},
		{"9d7f3e7d", 494878333, true, 4},
		{"7bbd", 15293, true, 2},
		{"25", 37, true, 1},
		{"4025", 37, false, 2},
	}
	for _, tc := range cases {
		raw := mustHex(t, tc.enc)
		got, err := ReadVarint(bufio.NewReader(bytes.NewReader(raw)))
		if err != nil || got != tc.v {
			t.Fatalf("ReadVarint(%s) = %d, %v; want %d", tc.enc, got, err, tc.v)
		}
		if !tc.minimal {
			continue
		}
		if enc := AppendVarint(nil, tc.v); !bytes.Equal(enc, raw) {
			t.Fatalf("AppendVarint(%d) = %x; want %s", tc.v, enc, tc.enc)
		}
		if n := VarintLen(tc.v); n != tc.encLenIs {
			t.Fatalf("VarintLen(%d) = %d; want %d", tc.v, n, tc.encLenIs)
		}
	}
}

func TestVarintBoundaries(t *testing.T) {
	for _, v := range []uint64{0, 63, 64, 16383, 16384, 1073741823, 1073741824, 1 << 40} {
		enc := AppendVarint(nil, v)
		if len(enc) != VarintLen(v) {
			t.Fatalf("AppendVarint(%d) wrote %d bytes; VarintLen says %d", v, len(enc), VarintLen(v))
		}
		got, err := ReadVarint(bufio.NewReader(bytes.NewReader(enc)))
		if err != nil || got != v {
			t.Fatalf("varint round-trip %d: got %d, %v", v, got, err)
		}
	}
}

// Shared fixture values, chosen so every byte position differs from its
// neighbours and a transposed field shows up as a diff.
const (
	fxTick    = 1024
	fxSeq     = 9
	fxTerrain = 0x0123456789abcdef
	fxWH      = 0xfeedfacecafebeef
	fxCTMS    = 1700000000000 // 0x0000018bcfe56800
)

func TestFrameGoldens(t *testing.T) {
	cases := []struct {
		name string
		enc  []byte
		want string
	}{
		{
			// proto 4, a client that has nothing yet (seq -1, tick 0).
			name: "hello",
			enc:  EncodeHello(Hello{Proto: Proto, SinceSeq: -1, SinceTick: 0, Ticket: "tkt"}),
			want: "18 01" +
				"0400" +
				"ffffffffffffffff" +
				"0000000000000000" +
				"0300" + "746b74",
		},
		{
			// move_to: dog id then node index, the sim's payload.
			name: "intent",
			enc: EncodeIntent(Intent{
				ID: 0x0102030405060708, Kind: evMoveTo,
				Payload: mustHex(t, "8877665544332211 0700"),
			}),
			want: "17 02" +
				"0807060504030201" +
				"0400" +
				"0a00" + "88776655443322110700",
		},
		{
			name: "resync",
			enc:  EncodeResync(Resync{HaveSeq: 4242}),
			want: "09 03" + "9210000000000000",
		},
		{
			name: "welcome",
			enc: EncodeWelcome(Welcome{
				Epoch: 1, Seq: 7, Tick: fxTick, Hz: 24,
				Role: RolePlayer, Terrain: fxTerrain, Park: "park-mythra",
			}),
			want: "2e 10" +
				"01000000" +
				"0700000000000000" +
				"0004000000000000" +
				"18000000" +
				"01" +
				"efcdab8967452301" +
				"0b" + "7061726b2d6d7974687261",
		},
		{
			// check_in by the dog that asked for it: intent id echoes back
			// so the sender can retire its pending entry.
			name: "event",
			enc: EncodeEvent(Event{
				Seq: fxSeq, Tick: fxTick, Kind: evCheckIn, Intent: 5,
				Payload: mustHex(t, "8877665544332211"),
			}),
			want: "25 11" +
				"0900000000000000" +
				"0004000000000000" +
				"0300" +
				"0500000000000000" +
				"0800" + "8877665544332211",
		},
		{
			name: "reject",
			enc:  EncodeReject(Reject{Intent: 5, Reason: 101}),
			want: "0d 12" + "0500000000000000" + "65000000",
		},
		{
			name: "snapshot",
			enc: EncodeSnapshot(Snapshot{
				Seq: fxSeq, Tick: fxTick, Epoch: 1, WH: fxWH,
				Terrain: fxTerrain, Z: mustHex(t, "deadbeef"),
			}),
			want: "2d 13" +
				"0900000000000000" +
				"0004000000000000" +
				"01000000" +
				"efbefecacefaedfe" +
				"efcdab8967452301" +
				"04000000" + "deadbeef",
		},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			want := mustHex(t, tc.want)
			if !bytes.Equal(tc.enc, want) {
				t.Fatalf("golden mismatch\n got %x\nwant %x", tc.enc, want)
			}
		})
	}
}

func TestDatagramGoldens(t *testing.T) {
	check := EncodeCheck(Check{Tick: fxTick, WH: fxWH, CTMS: fxCTMS})
	wantCheck := mustHex(t, "01"+
		"0004000000000000"+
		"efbefecacefaedfe"+
		"0068e5cf8b010000")
	if !bytes.Equal(check, wantCheck) {
		t.Fatalf("check golden mismatch\n got %x\nwant %x", check, wantCheck)
	}
	if len(check) != CheckLen {
		t.Fatalf("check is %d bytes; the layout says %d", len(check), CheckLen)
	}

	verdict := EncodeVerdict(Verdict{
		Tick: fxTick, Now: 1030, CTMS: fxCTMS,
		Flags: VerdictKnown | VerdictOK,
		CW:    ModuleWord("9abcdef0"), PW: ModuleWord("12345678"),
	})
	// cw/pw land on the wire as the hash bytes in display order.
	wantVerdict := mustHex(t, "02"+
		"0004000000000000"+
		"0604000000000000"+
		"0068e5cf8b010000"+
		"03"+
		"9abcdef0"+
		"12345678")
	if !bytes.Equal(verdict, wantVerdict) {
		t.Fatalf("verdict golden mismatch\n got %x\nwant %x", verdict, wantVerdict)
	}
	if len(verdict) != VerdictLen {
		t.Fatalf("verdict is %d bytes; the layout says %d", len(verdict), VerdictLen)
	}
}

func TestModuleWord(t *testing.T) {
	// The display hash is hex of sha256[:4]; the u32 is those bytes read
	// little-endian, so re-encoding reproduces the display order exactly.
	if got := ModuleWord("9abcdef0"); got != 0xf0debc9a {
		t.Fatalf("ModuleWord(9abcdef0) = %08x; want f0debc9a", got)
	}
	if got := HexWord(ModuleWord("9abcdef0")); got != "9abcdef0" {
		t.Fatalf("HexWord round-trip = %q; want 9abcdef0", got)
	}
	if got := ModuleWord(""); got != 0 {
		t.Fatalf("ModuleWord of an unloaded module = %08x; want 0", got)
	}
	if got := ModuleWord("zzzzzzzz"); got != 0 {
		t.Fatalf("ModuleWord of a non-hex string = %08x; want 0", got)
	}
}

// Each encoder declares its payload length before writing the fields, so
// the two can drift. Nothing downstream notices a length that is merely
// short — the next frame just starts inside this one — which makes this
// worth checking directly.
func TestLengthPrefixMatchesBody(t *testing.T) {
	frames := [][]byte{
		EncodeHello(Hello{Proto: Proto, Ticket: "ticket"}),
		EncodeIntent(Intent{Payload: []byte{1, 2, 3}}),
		EncodeResync(Resync{HaveSeq: 1}),
		EncodeWelcome(Welcome{Park: "park-mythra"}),
		EncodeEvent(Event{Payload: []byte{1}}),
		EncodeReject(Reject{}),
		EncodeSnapshot(Snapshot{Z: []byte{1, 2, 3, 4, 5}}),
		EncodeSnapshot(Snapshot{Z: bytes.Repeat([]byte{0}, 70000)}),
	}
	for _, f := range frames {
		body, err := ReadVarint(bufio.NewReader(bytes.NewReader(f)))
		if err != nil {
			t.Fatalf("varint: %v", err)
		}
		kind := f[VarintLen(body)]
		if want := uint64(len(f)) - uint64(VarintLen(body)); body != want {
			t.Fatalf("kind %d declares a %d-byte body but wrote %d", kind, body, want)
		}
		if got := FrameSize(len(f) - VarintLen(body) - 1); got != len(f) {
			t.Fatalf("kind %d: FrameSize says %d, frame is %d", kind, got, len(f))
		}
	}
}

func TestFrameRoundTrip(t *testing.T) {
	// One instance of every kind, written into a stream and read back
	// through the decoder the session uses.
	hello := Hello{Proto: Proto, SinceSeq: -1, Ticket: "a-ticket"}
	intent := Intent{ID: 7, Kind: evBoostSet, Payload: []byte{1, 2, 3, 4, 5, 6, 7, 8, 1}}
	resync := Resync{HaveSeq: 12345}
	welcome := Welcome{Epoch: 3, Seq: fxSeq, Tick: fxTick, Hz: 24, Role: RoleSpectator, Terrain: fxTerrain, Park: "park-mythra"}
	event := Event{Seq: fxSeq, Tick: fxTick, Kind: evJoin, Intent: 11, Payload: []byte{9, 9}}
	reject := Reject{Intent: 11, Reason: 100}
	snapshot := Snapshot{Seq: fxSeq, Tick: fxTick, Epoch: 3, WH: fxWH, Terrain: fxTerrain, Z: bytes.Repeat([]byte{0xa5}, 40000)}

	var stream bytes.Buffer
	for _, b := range [][]byte{
		EncodeHello(hello), EncodeIntent(intent), EncodeResync(resync),
		EncodeWelcome(welcome), EncodeEvent(event), EncodeReject(reject),
		EncodeSnapshot(snapshot),
	} {
		stream.Write(b)
	}

	fr := NewReader(&stream)
	type decoded struct {
		kind byte
		val  any
	}
	var got []decoded
	for {
		kind, payload, err := fr.Next()
		if errors.Is(err, io.EOF) {
			break
		}
		if err != nil {
			t.Fatalf("frame read: %v", err)
		}
		var v any
		switch kind {
		case KindHello:
			v, err = DecodeHello(payload)
		case KindIntent:
			v, err = DecodeIntent(payload)
		case KindResync:
			v, err = DecodeResync(payload)
		case KindWelcome:
			v, err = DecodeWelcome(payload)
		case KindEvent:
			v, err = DecodeEvent(payload)
		case KindReject:
			v, err = DecodeReject(payload)
		case KindSnapshot:
			v, err = DecodeSnapshot(payload)
		default:
			t.Fatalf("unknown frame kind %d", kind)
		}
		if err != nil {
			t.Fatalf("decode kind %d: %v", kind, err)
		}
		got = append(got, decoded{kind, v})
	}

	want := []decoded{
		{KindHello, hello}, {KindIntent, intent}, {KindResync, resync},
		{KindWelcome, welcome}, {KindEvent, event}, {KindReject, reject},
		{KindSnapshot, snapshot},
	}
	if len(got) != len(want) {
		t.Fatalf("read %d frames; wrote %d", len(got), len(want))
	}
	for i := range want {
		if got[i].kind != want[i].kind || !reflect.DeepEqual(got[i].val, want[i].val) {
			t.Fatalf("frame %d round-trip\n got %d %#v\nwant %d %#v",
				i, got[i].kind, got[i].val, want[i].kind, want[i].val)
		}
	}
}

func TestDatagramRoundTrip(t *testing.T) {
	chk := Check{Tick: fxTick, WH: fxWH, CTMS: fxCTMS}
	if got, err := DecodeCheck(EncodeCheck(chk)); err != nil || got != chk {
		t.Fatalf("check round-trip: %#v, %v", got, err)
	}
	for _, flags := range []uint8{0, VerdictKnown, VerdictKnown | VerdictOK} {
		v := Verdict{Tick: fxTick, Now: fxTick + 6, CTMS: fxCTMS, Flags: flags, CW: 1, PW: 2}
		if got, err := DecodeVerdict(EncodeVerdict(v)); err != nil || got != v {
			t.Fatalf("verdict round-trip: %#v, %v", got, err)
		}
	}
}

func TestDecodeRejectsMalformed(t *testing.T) {
	full := EncodeEvent(Event{Seq: 1, Tick: 2, Kind: evJoin, Intent: 3, Payload: []byte{4}})
	payload := full[2:] // past the varint and kind byte

	if _, err := DecodeEvent(payload[:len(payload)-1]); err == nil {
		t.Fatal("truncated event payload decoded")
	}
	if _, err := DecodeEvent(append(append([]byte{}, payload...), 0xff)); err == nil {
		t.Fatal("event payload with trailing bytes decoded")
	}
	// A length field larger than the bytes that follow must not be trusted.
	bad := EncodeSnapshot(Snapshot{Z: []byte{1, 2, 3, 4}})[2:]
	bad[39] = 0xff // top byte of zlen, at offset 36 of the payload
	if _, err := DecodeSnapshot(bad); err == nil {
		t.Fatal("snapshot with an overlong zlen decoded")
	}
	if _, err := DecodeHello(nil); err == nil {
		t.Fatal("empty hello payload decoded")
	}
	if _, err := DecodeCheck(EncodeVerdict(Verdict{})); err == nil {
		t.Fatal("verdict decoded as a check")
	}
	if _, err := DecodeCheck(EncodeCheck(Check{})[:24]); err == nil {
		t.Fatal("truncated check decoded")
	}
}

func TestReaderRejectsOversize(t *testing.T) {
	var b bytes.Buffer
	b.Write(AppendVarint(nil, MaxFrameLen+1))
	if _, _, err := NewReader(&b).Next(); !errors.Is(err, ErrFrameTooLarge) {
		t.Fatalf("oversized frame: got %v; want ErrFrameTooLarge", err)
	}

	b.Reset()
	b.Write(AppendVarint(nil, 0))
	if _, _, err := NewReader(&b).Next(); !errors.Is(err, ErrShortFrame) {
		t.Fatalf("empty frame body: got %v; want ErrShortFrame", err)
	}

	// A truncated body is a short read, not a decode: the peer went away
	// mid-frame and the session ends.
	b.Reset()
	b.Write(AppendVarint(nil, 16))
	b.Write([]byte{KindEvent, 1, 2, 3})
	if _, _, err := NewReader(&b).Next(); err == nil {
		t.Fatal("truncated frame body read as a whole frame")
	}
}

func TestRoleCode(t *testing.T) {
	if RoleCode("player") != RolePlayer || RoleCode("spectator") != RoleSpectator || RoleCode("") != RoleSpectator {
		t.Fatal("role encoding does not match the wire's 0 spectator / 1 player")
	}
}
